### ISSUE TITLE
RemoteWebDriver unconditionally wraps W3C WebDriver responses in value field

### ISSUE THREAD

**Issue by mehmetg** [SeleniumHQ/selenium#2731]
The [get element rect spec](http://w3c.github.io/webdriver/webdriver-spec.html#get-element-rect) is not clear on the structure of the model. We need agreement between bindings and the geckodriver. Issue link: https://github.com/mozilla/geckodriver/issues/207
## Meta -
OS:  MacOS 10.11.6

Selenium Version:  2.53.1/GeckoDriver 0.9.0
 
Browser:  FF48 w/ GeckoDriver 0.9.0

Browser Version:  48.0.2

## Expected Behavior -
Referenced code below expects:
A response like this:
```json
{  
   "point":null,
   "dimension":null,
   "width":98,
   "y":10954,
   "x":20,
   "hCode":-620756992,
   "height":17,
   "class":"org.openqa.selenium.Rectangle"
}
```
https://github.com/SeleniumHQ/selenium/blob/master/py/selenium/webdriver/remote/webelement.py#L393-L402
## Actual Behavior -
But gets this:
```json
{  
   "status":0,
   "sessionId":"b7040c41-00d7-4486-8c86-b8c715bdff2f",
   "value":{  
      "point":null,
      "dimension":null,
      "width":600,
      "y":314,
      "x":20,
      "hCode":44480,
      "height":87,
      "class":"org.openqa.selenium.Rectangle"
   },
   "state":"success",
   "hCode":493044042,
   "class":"org.openqa.selenium.remote.Response"
}
```

**Comment by andreastt** [comment]
See my explanation in https://github.com/mozilla/geckodriver/issues/207#issuecomment-245107791.

**Comment by shs96c** [comment]
Hmph! I can fix this.

**Comment by shs96c** [comment]
More specifically, I think that this is the perfect time to flesh out the response codecs, as there's obviously some work to be done. Interestingly, my local tests of both the python and java bindings pass. I suspect there's a problem with the remote webdriver tests that should have picked this up.

**Comment by mehmetg** [comment]
@shs96c: As far as I can tell the Java response decoder is masking the issue and sticking the response(rect data) into the value member of the response somehow.  As seen [here](https://github.com/SeleniumHQ/selenium/blob/eb37f619f04c14187071f442669e32a87140157e/java/client/src/org/openqa/selenium/remote/http/AbstractHttpResponseCodec.java). 
If the first attempt fails everything it the JSON seems to be assigned to the value field as long as there's a 2xx status code returned. This seems to start at line 97.

Java bindings are a bit too forgiving while decoding JSON and they may be masking more issues. :)

I have no Se dev setup ready so, it was a code read through mostly. Excuse me, if I am off the mark with my comment.

**Comment by shs96c** [comment]
Picking up this one. I think there are interesting cases to consider:

1. W3C local end to W3C remote end --- no changes needed
2. OSS local end to OSS remote end --- no changes needed
3. W3C local end to OSS remote end --- converting commands and responses needed
4. OSS local end to W3C remote end --- converting commands and responses needed

We can already transform outgoing commands, so "all" we need to do now is the opposite end. Fortunately, the `RemoteWebDriver` internals have a protocol independent representation of the data structures. "How hard can this be"?

**Comment by shs96c** [comment]
There is another quirk, in that the remote webdriver server needs to connects to two other nodes (normally an end node and local end), and both of those ends may speak different dialects.

**Comment by shs96c** [comment]
Fixed with 0dadb0092257a307e8423a42a7d7c86300e87874
